### PR TITLE
Initial Extending Rules Documentation

### PR DESCRIPTION
**Pull request by stevepurkiss**

### CODE REVIEW

File: extending_rules/providing_context_to_rules_plugins.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# Providing Context to Rules Plugins
+
+To provide context to a Rules Plugin, for example an entity for the entity delete action plugin, use the following format:
+
+```
+ *   context = {
+ *     "entity" = @ContextDefinition("entity",
+ *       label = @Translation("Entity"),
+ *       description = @Translation("Specifies the entity, which should be deleted permanently.")
+ *     )
```
Comment (fubhy) [inline extending_rules/providing_context_to_rules_plugins.md]: Misses the closing curly bracket.

File: extending_rules/rules_action_plugins.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# Rules Action Plugins
+
+To implement a Rules Action Plugin, place your plugin code in ```\Drupal\{module_name}\Plugin\Action\```, for example:
```
Comment (fubhy) [inline extending_rules/rules_action_plugins.md]: "Place your plugin code under the namespace ..."

File: extending_rules/rules_condition_plugins.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# Rules Condition Plugins
+
+To implement a Rules Condition Plugin, place your plugin code in ```\Drupal\{module_name}\Plugin\Condition\```, for example:
```
Comment (fubhy) [inline extending_rules/rules_condition_plugins.md]: Same here (yeah I know I sai to only add the namespace but I think it's good to have both).

File: extending_rules/rules_data_processor_plugins.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
+# Rules Data Processor Plugins
+
+Data processors are for processing the values resulting from the configured data selection - Rules implements a processor for applying offsets to date values as well as a custom-PHP-code processor.
+
+To implement a Rules Data Processor Plugin, place your plugin code in ```\Drupal\{module_name}\Plugin\DataProcessor\```, for example:
```
Comment (fubhy) [inline extending_rules/rules_data_processor_plugins.md]: Same.

File: extending_rules/rules_data_processor_plugins.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
+# Rules Data Processor Plugins
+
+Data processors are for processing the values resulting from the configured data selection - Rules implements a processor for applying offsets to date values as well as a custom-PHP-code processor.
```
Comment (fubhy) [inline extending_rules/rules_data_processor_plugins.md]: custom PHP code. Why the dashes? :)

File: extending_rules/rules_expression_plugins.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# Rules Expression Plugins
+
+To implement a Rules Expression Plugin, place your plugin code in ```\Drupal\{module_name}\Plugin\Expression\```, for example:
```
Comment (fubhy) [inline extending_rules/rules_expression_plugins.md]: Same.

File: extending_rules/rules_expression_plugins.md [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+
+```some example code goes here```
```
Comment (fubhy) [inline extending_rules/rules_expression_plugins.md]: Let's not add todo-ish stuff here.

**Comment by stevepurkiss** [comment]
Changing DataProcessor back to RulesDataProcessor & Expression back to RulesExpression as per https://www.drupal.org/node/2471533

**Comment by stevepurkiss** [comment]
Changing DataProcessor back to RulesDataProcessor & Expression back to RulesExpression as per https://www.drupal.org/node/2471533

**Comment by fubhy** [comment]
@stevepurkiss Thanks :+1:

**Pull request merged**